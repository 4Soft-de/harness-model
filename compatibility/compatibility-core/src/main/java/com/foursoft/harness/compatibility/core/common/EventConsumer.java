/*-
 * ========================LICENSE_START=================================
 * Compatibility Core
 * %%
 * Copyright (C) 2020 - 2023 4Soft GmbH
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * =========================LICENSE_END==================================
 */
package com.foursoft.harness.compatibility.core.common;

import jakarta.xml.bind.ValidationEvent;
import jakarta.xml.bind.ValidationEventLocator;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.w3c.dom.Node;

import java.util.function.Consumer;

/**
 * Simple {@link Consumer} for {@link ValidationEvent} which logs warnings / errors with a {@link Logger}.
 */
public class EventConsumer implements Consumer<ValidationEvent> {

    private static final Logger LOGGER = LoggerFactory.getLogger(EventConsumer.class);

    @Override
    public void accept(final ValidationEvent t) {
        if (t.getSeverity() == ValidationEvent.WARNING) {
            handleWarning(t);
        } else {
            handleError(t);
        }
    }

    private static String getLocation(final ValidationEventLocator locator) {
        final StringBuilder b = new StringBuilder(100);
        final int lineNumber = locator.getLineNumber();
        if (lineNumber > 0) {
            final int columnNumber = locator.getColumnNumber();
            b.append(locator)
                    .append(':')
                    .append(columnNumber > -1 ? columnNumber : "?");
        }

        final Node n = locator.getNode();
        if (n != null) {
            final Node id = n.getAttributes().getNamedItem("id");
            if (id != null) {
                b.append(" (").append(id).append(')');
            }
        }

        if (b.length() > 0) {
            b.append(": ");
        }
        return b.toString();
    }

    private static void handleWarning(final ValidationEvent e) {
        LOGGER.warn("{}{}", getLocation(e.getLocator()), e.getMessage(), e.getLinkedException());
    }

    private static void handleError(final ValidationEvent e) {
        LOGGER.error("{}{}", getLocation(e.getLocator()), e.getMessage(), e.getLinkedException());
    }

}
