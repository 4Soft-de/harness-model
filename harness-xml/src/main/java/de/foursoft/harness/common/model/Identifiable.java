package de.foursoft.harness.common.model;

public interface Identifiable {

    String getXmlId();

}
