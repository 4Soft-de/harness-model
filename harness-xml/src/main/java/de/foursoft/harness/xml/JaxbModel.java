/*-
 * ========================LICENSE_START=================================
 * xml-runtime
 * %%
 * Copyright (C) 2019 4Soft GmbH
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * =========================LICENSE_END==================================
 */
package de.foursoft.harness.xml;

public class JaxbModel<R, I> {

    private final R rootElement;
    private final IdLookupProvider<I> idLookup;

    public JaxbModel(final R rootElement, final IdLookupProvider<I> idLookupProvider) {
        this.rootElement = rootElement;
        this.idLookup = idLookupProvider;
    }

    /**
     * Copy constructor for use in inheriting classes. This can be used if the
     * {@link JaxbModel} returned by the {@link ExtendedUnmarshaller} shall be
     * extended with its own convenience methods.
     * 
     * @param model
     */
    protected JaxbModel(final JaxbModel<R, I> model) {
        this.rootElement = model.rootElement;
        this.idLookup = model.idLookup;
    }

    public R getRootElement() {
        return rootElement;
    }

    /**
     * A {@link IdLookupProvider} to allow the lookup of identifiable elements
     * in this model.
     * <p/>
     * The {@link IdLookupProvider} must be initialized during the unmarshalling
     * process (see
     * {@link ExtendedUnmarshaller#withIdMapper(Class, java.util.function.Function)}).
     * 
     * @return
     */
    public IdLookupProvider<I> getIdLookup() {
        if (idLookup == null) {
            throw new JaxbModelException("idLookup can not be used. It was not initialized during the unmarshalling.");
        }
        return idLookup;
    }

}
