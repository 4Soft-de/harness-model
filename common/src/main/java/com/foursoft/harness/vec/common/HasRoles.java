/*-
 * ========================LICENSE_START=================================
 * vec-common
 * %%
 * Copyright (C) 2020 - 2022 4Soft GmbH
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * =========================LICENSE_END==================================
 */
package com.foursoft.harness.vec.common;

import com.foursoft.harness.vec.common.util.DelegationUtils;
import com.foursoft.harness.vec.common.util.StreamUtils;

import java.util.List;
import java.util.Optional;

@FunctionalInterface
public interface HasRoles<X> {

    List<X> getRoles();

    default <T extends X> List<T> getRolesWithType(final Class<T> type) {
        return DelegationUtils.getFromListWithType(getRoles(), type);
    }

    /**
     * Gets the first role with the given type.
     * <b>Warning: There might be multiple roles with the given type.
     * Only use this method if you are sure there will just be one element!</b>
     *
     * @param type Class of T.
     * @param <T>  Type of role to filter for.
     * @return The first role with the given type if found, else an empty optional.
     */
    default <T extends X> Optional<T> getRoleWithType(final Class<T> type) {
        return DelegationUtils.getFromListWithTypeAsStream(getRoles(), type)
                .collect(StreamUtils.findOneOrNone());
    }

}
