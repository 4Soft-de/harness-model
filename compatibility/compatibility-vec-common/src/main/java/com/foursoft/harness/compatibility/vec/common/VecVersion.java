/*-
 * ========================LICENSE_START=================================
 * Compatibility VEC Common
 * %%
 * Copyright (C) 2020 - 2023 4Soft GmbH
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * =========================LICENSE_END==================================
 */
package com.foursoft.harness.compatibility.vec.common;

import com.foursoft.harness.vec.common.HasVecVersion;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Arrays;
import java.util.Optional;
import java.util.regex.Pattern;

/**
 * Enumeration containing all supported VEC Versions.
 */
public enum VecVersion {

    // The VEC versions should be added in order of their version, canBeConvertedTo depends on it.

    /**
     * Enum value for the VEC 1.1.X.
     */
    VEC11X("1.1.X", "1.1.3"),

    /**
     * Enum value for the VEC 1.2.X.
     */
    VEC12X("1.2.X", "1.2.0");

    private static final Logger LOGGER = LoggerFactory.getLogger(VecVersion.class);

    private static final String VEC_VERSION_XML_FORMAT = "<VecVersion>%s</VecVersion>";

    private final String identifier;
    private final Pattern marker;
    private final String currentVersion;

    VecVersion(final String identifier, final String currentVersion) {
        this.identifier = identifier;
        this.currentVersion = currentVersion;
        marker = createMarkerPattern(identifier);
    }

    /**
     * Returns the actual current version of the enum entry (eg. 1.1.3 for 1.1.X).
     *
     * @return The actual current version of the enum entry.
     */
    public String getCurrentVersion() {
        return currentVersion;
    }

    /**
     * Returns the identifier of the enum entry. Shouldn't be used for writing it to a file.
     *
     * @return The identifier of the enum entry.
     */
    public String getIdentifier() {
        return identifier;
    }

    /**
     * Tries to find the version within the given String.
     *
     * @param input String which should contain a version.
     * @return Possibly-empty Optional containing the found version.
     */
    public static Optional<VecVersion> findVersion(final String input) {
        final Optional<VecVersion> vecVersion = Arrays.stream(VecVersion.values())
                .filter(c -> c.marker.matcher(input).find())
                .findFirst();
        if (vecVersion.isEmpty()) {
            LOGGER.debug("No VecVersion could be determined by the given input: {}", input);
        }
        return vecVersion;
    }

    /**
     * Tries to find the version for the given {@code VecContent}.
     *
     * @param value VecContent which should have a version.
     * @return Possibly-empty Optional containing the found version.
     */
    public static Optional<VecVersion> findVersion(final HasVecVersion value) {
        final String vecVersionFormatted = String.format(VEC_VERSION_XML_FORMAT, value.getVecVersion());
        return findVersion(vecVersionFormatted);
    }

    /**
     * Method telling whether this Version can be converted to the given version or not.
     * <p>
     * This will most likely return {@code true} if there is only one minor or major version in between.
     *
     * @param otherVersion Version to check it this version can be converted to.
     * @return {@code true} if this version can be converted to the given version, else {@code false}.
     */
    public boolean canBeConvertedTo(final VecVersion otherVersion) {
        final int thisPosition = ordinal();
        final int otherPosition = otherVersion.ordinal();
        return Math.abs(thisPosition - otherPosition) == 1;
    }

    @Override
    public String toString() {
        return "VecVersion{" +
                "identifier='" + identifier + '\'' +
                ", currentVersion='" + currentVersion + '\'' +
                '}';
    }

    private static Pattern createMarkerPattern(final String version) {
        final String versionAsRegex = version
                .replace("X", "\\d+")  // In theory an X could be replaced with a double-digit value.
                .replace(".", "\\.");  // Dots are special RegEx characters, need to be escaped.

        // Slashes are also special RegEx characters.
        final String fullStringToFind = String.format(VEC_VERSION_XML_FORMAT, versionAsRegex)
                .replace("/", "\\/");  // Slashes are special RegEx characters, need to be escaped.
        return Pattern.compile(fullStringToFind);
    }

}
