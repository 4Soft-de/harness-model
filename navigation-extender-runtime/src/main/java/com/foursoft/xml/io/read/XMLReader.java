/*-
 * ========================LICENSE_START=================================
 * navigation-extender-runtime
 * %%
 * Copyright (C) 2019 - 2020 4Soft GmbH
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * =========================LICENSE_END==================================
 */
package com.foursoft.xml.io.read;

import com.foursoft.xml.ExtendedUnmarshaller;
import com.foursoft.xml.JaxbModel;
import com.foursoft.xml.io.utils.ValidationEventLogger;
import com.foursoft.xml.io.utils.XMLIOException;
import jakarta.xml.bind.JAXBException;
import jakarta.xml.bind.ValidationEvent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Objects;
import java.util.function.Consumer;
import java.util.function.Function;

/**
 * A generic xml reader which includes the back references, id mapper and a simple event log.
 *
 * @param <T> The root class to deserialize
 * @param <I> The class to identify xml objects
 */
public class XMLReader<T, I> {
    private static final Logger LOGGER = LoggerFactory.getLogger(XMLReader.class);
    private final ExtendedUnmarshaller<T, I> unmarshaller;

    /**
     * It will log warnings, errors and fatal validation events to the slf4j, see @{@link ValidationEventLogger}
     *
     * @param rootTypeClass     the root class of the jaxb model
     * @param identifiableClass the class of the identifiable element
     * @param idMapper          the mapper function to get the id of each element
     * @see ExtendedUnmarshaller
     */
    public XMLReader(final Class<T> rootTypeClass,
                     final Class<I> identifiableClass,
                     final Function<I, String> idMapper) {
        this(rootTypeClass, identifiableClass, idMapper, new ValidationEventLogger());
    }

    /**
     * @param rootTypeClass           the root class of the jaxb model
     * @param identifiableClass       the class of the identifiable element
     * @param idMapper                the mapper function to get the id of each element
     * @param validationEventConsumer a consumer for validation events
     * @see ExtendedUnmarshaller
     */
    public XMLReader(final Class<T> rootTypeClass,
                     final Class<I> identifiableClass,
                     final Function<I, String> idMapper,
                     final Consumer<ValidationEvent> validationEventConsumer) {
        try {
            unmarshaller = new ExtendedUnmarshaller<T, I>(rootTypeClass)
                    .withBackReferences()
                    .withIdMapper(identifiableClass, idMapper)
                    .withEventLogging(validationEventConsumer);
        } catch (final Exception e) {
            throw new XMLIOException("Cannot initialize unmarshaller.", e);
        }
    }

    /**
     * Builds the complete JAXB tree structure of an xml file.
     * Uses the already registered event consumer for validation events
     *
     * @param filename the path name of the XML file.
     * @return the JAXB object structure representing the xml file
     * @throws XMLIOException in case of an IOException
     */
    public T read(final String filename) {
        Objects.requireNonNull(filename);
        final Path path = Paths.get(filename).toAbsolutePath();
        if (!Files.exists(path)) {
            throw new XMLIOException("Given path: " + path + " doesn't exists.");
        }
        if (!Files.isReadable(path)) {
            throw new XMLIOException("Given path: " + path + " isn't readable.");
        }
        try (final InputStream is = new BufferedInputStream(Files.newInputStream(path))) {
            return read(is);
        } catch (final IOException e) {
            throw new XMLIOException("Error reading file " + filename, e);
        }
    }

    /**
     * Builds the complete JAXB tree structure of an xml stream.
     * Uses the already registered event consumer for validation events
     *
     * @param inputStream the input stream to read from
     * @return the JAXB object structure representing the xml file.
     * @throws XMLIOException in case of a JAXBException
     */
    public T read(final InputStream inputStream) {
        Objects.requireNonNull(inputStream);
        return readModel(inputStream).getRootElement();
    }

    /**
     * Builds the complete JAXB tree structure of an xml stream.
     * Uses the already registered event consumer for validation events
     *
     * @param inputStream the input stream to read from
     * @return the JAXB object structure representing the xml file.
     * @throws XMLIOException in case of a JAXBException
     */
    public JaxbModel<T, I> readModel(final InputStream inputStream) {
        Objects.requireNonNull(inputStream);
        try {
            final long start = System.currentTimeMillis();
            final JaxbModel<T, I> result = unmarshaller.unmarshall(inputStream);
            LOGGER.trace("Finished loading XML. Took {} ms", System.currentTimeMillis() - start);
            return result;
        } catch (final JAXBException e) {
            throw new XMLIOException(e.getMessage(), e);
        }
    }
}


