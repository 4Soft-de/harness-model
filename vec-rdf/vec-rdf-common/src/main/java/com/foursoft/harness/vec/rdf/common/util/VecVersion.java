/*-
 * ========================LICENSE_START=================================
 * VEC RDF Common
 * %%
 * Copyright (C) 2024 4Soft GmbH
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * =========================LICENSE_END==================================
 */
package com.foursoft.harness.vec.rdf.common.util;

import com.foursoft.harness.navext.runtime.io.read.XMLReader;
import com.foursoft.harness.navext.runtime.io.write.XMLWriter;
import com.foursoft.harness.navext.runtime.model.Identifiable;
import com.foursoft.harness.vec.rdf.common.exception.VecRdfException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.w3c.dom.Document;
import org.w3c.dom.NodeList;

import java.util.Arrays;
import java.util.Objects;

public enum VecVersion {

    V113("1.1.3"), V120("1.2.0"), V121("1.2.1"), V122("1.2.2"), V200("2.0.0"), V201("2.0.1"), V202("2.0.2"), V210(
            "2.1.0");

    private static final Logger LOGGER = LoggerFactory.getLogger(VecVersion.class);

    private final String version;

    VecVersion(String version) {
        this.version = version;
    }

    public String getVersion() {
        return version;
    }

    public XMLReader<? extends Identifiable, Identifiable> resolveReader() {
        if ("1.1.3".equals(version)) {
            return new com.foursoft.harness.vec.v113.VecReader();
        }
        if (version.startsWith("1.2.")) {
            return new com.foursoft.harness.vec.v12x.VecReader();
        }
        if (version.startsWith("2.")) {
            return new com.foursoft.harness.vec.v2x.VecReader();
        }
        throw new VecRdfException("No VEC Api for Version: '" + version + "' found.");
    }

    public XMLWriter<? extends Identifiable> resolveWriter() {
        if ("1.1.3".equals(version)) {
            return new com.foursoft.harness.vec.v113.VecWriter();
        }
        if (version.startsWith("1.2.")) {
            return new com.foursoft.harness.vec.v12x.VecWriter();
        }
        if (version.startsWith("2.")) {
            return new com.foursoft.harness.vec.v2x.VecWriter();
        }
        throw new VecRdfException("No VEC Api for Version: '" + version + "' found.");
    }

    public static VecVersion guessVersion(Document document) {
        LOGGER.debug("Trying to guess VEC version.");
        NodeList versions = document.getElementsByTagName("VecVersion");
        if (versions.getLength() == 0) {
            throw new VecRdfException("Unable to find 'VecVersion' tag in.");
        }
        String version = versions.item(0)
                .getTextContent();
        LOGGER.info("VEC version detected: {}", version);
        return VecVersion.fromVersion(version);
    }

    public static VecVersion fromVersion(String version) {
        Objects.requireNonNull(version, "version must not be null");
        VecVersion[] enumConstants = VecVersion.class.getEnumConstants();
        return Arrays.stream(enumConstants)
                .filter(v -> version.equals(v.getVersion()))
                .findFirst()
                .orElseThrow(() -> new VecRdfException("Unsupported VEC Version: " + version));
    }
}
