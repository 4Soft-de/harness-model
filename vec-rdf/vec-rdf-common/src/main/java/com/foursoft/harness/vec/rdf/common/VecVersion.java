/*-
 * ========================LICENSE_START=================================
 * VEC RDF Common
 * %%
 * Copyright (C) 2024 4Soft GmbH
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * =========================LICENSE_END==================================
 */
package com.foursoft.harness.vec.rdf.common;

import com.foursoft.harness.vec.rdf.common.exception.VecRdfException;

import java.io.InputStream;
import java.util.Arrays;
import java.util.Objects;

public enum VecVersion {
    V113("1.1.3", Packages.V113, true),
    V120("1.2.0", Packages.V12X, false),
    V121("1.2.1", Packages.V12X, false),
    V122("1.2.2", Packages.V12X, true),
    V200("2.0.0", Packages.V2X, false),
    V201("2.0.1", Packages.V2X, false),
    V202("2.0.2", Packages.V2X, false),
    V210("2.1.0", Packages.V2X, true);

    private final static String modelLocationPattern = "/vec/v%1$s/vec-%1$s.mdxml";
    private final String versionString;
    private final String apiPackage;
    private final boolean latest;

    VecVersion(final String versionString, final String apiPackage,
               final boolean latest) {
        this.versionString = versionString;
        this.apiPackage = apiPackage;
        this.latest = latest;
    }

    public String getVersionString() {
        return versionString;
    }

    public String getApiPackage() {
        return apiPackage;
    }

    public boolean isLatest() {
        return latest;
    }

    public InputStream getModelInputStream() {
        final String modelLocation = String.format(modelLocationPattern, versionString);
        final InputStream resourceAsStream = this.getClass().getResourceAsStream(modelLocation);
        return resourceAsStream;
    }

    public static VecVersion findLatestVersionForApiPackage(final String apiPackage) {
        return Arrays.stream(values())
                .filter(literal -> literal.apiPackage.equals(apiPackage))
                .filter(VecVersion::isLatest)
                .findAny()
                .orElseThrow(() -> new VecRdfException("No UML model found for VEC API package: " + apiPackage));
    }

    public static VecVersion fromVersionString(String version) {
        Objects.requireNonNull(version, "version must not be null");
        VecVersion[] enumConstants = VecVersion.class.getEnumConstants();
        return Arrays.stream(enumConstants)
                .filter(v -> version.equals(v.getVersionString()))
                .findFirst()
                .orElseThrow(() -> new VecRdfException("Unsupported VEC Version: " + version));
    }

    private final static class Packages {
        public static final String V113 = "com.foursoft.harness.vec.v113";
        public static final String V12X = "com.foursoft.harness.vec.v12x";
        public static final String V2X = "com.foursoft.harness.vec.v2x";
    }

}
