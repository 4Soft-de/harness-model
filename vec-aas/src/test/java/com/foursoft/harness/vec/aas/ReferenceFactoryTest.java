package com.foursoft.harness.vec.aas;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.foursoft.harness.navext.runtime.JaxbModel;
import com.foursoft.harness.navext.runtime.io.read.XMLReader;
import com.foursoft.harness.navext.runtime.model.Identifiable;
import com.foursoft.harness.vec.rdf.common.VecVersion;
import com.foursoft.harness.vec.v2x.VecWireElement;
import org.eclipse.digitaltwin.aas4j.v3.model.Reference;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.w3c.dom.Document;

import java.io.InputStream;

import static com.foursoft.harness.vec.rdf.common.util.VecXmlApiUtils.*;

@ExtendWith(SnapshotExtension.class)
class ReferenceFactoryTest {

    Expect expect;

    @Test
    void test() {
        final InputStream inputFile = this.getClass().getResourceAsStream("/coroflex-ti-9-2611-35.vec");
        final Document document = loadDocument(inputFile);

        final VecVersion version = guessVersion(document);

        final XMLReader<?, Identifiable> reader = resolveReader(version);

        final JaxbModel<?, Identifiable> jaxbModel = reader.readModel(document);

        final ReferenceFactory factory = new ReferenceFactory(new AasNamingStrategy(), "https://www.test.com",
                                                              (Identifiable) jaxbModel.getRootElement());

        final VecWireElement wireElement = jaxbModel.getIdLookup().findById(VecWireElement.class,
                                                                            "WireElement_00064").orElseThrow();

        final Reference reference = factory.localReferenceFor(wireElement);

        expect.serializer("json").toMatchSnapshot(reference);

    }

}