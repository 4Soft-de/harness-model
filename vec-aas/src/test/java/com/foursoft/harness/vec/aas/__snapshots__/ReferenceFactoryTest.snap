com.foursoft.harness.vec.aas.ReferenceFactoryTest.test=[
  {
    "keys": [
      {
        "type": "GLOBAL_REFERENCE",
        "value": "https://www.test.com"
      },
      {
        "type": "FRAGMENT_REFERENCE",
        "value": "Content_00000"
      },
      {
        "type": "FRAGMENT_REFERENCE",
        "value": "documentVersions"
      },
      {
        "type": "FRAGMENT_REFERENCE",
        "value": "DocumentVersion_00006"
      },
      {
        "type": "FRAGMENT_REFERENCE",
        "value": "specifications"
      },
      {
        "type": "FRAGMENT_REFERENCE",
        "value": "WireSpecification_00061"
      },
      {
        "type": "FRAGMENT_REFERENCE",
        "value": "wireElement"
      },
      {
        "type": "FRAGMENT_REFERENCE",
        "value": "WireElement_00062"
      },
      {
        "type": "FRAGMENT_REFERENCE",
        "value": "subWireElements"
      },
      {
        "type": "FRAGMENT_REFERENCE",
        "value": "WireElement_00063"
      },
      {
        "type": "FRAGMENT_REFERENCE",
        "value": "subWireElements"
      },
      {
        "type": "FRAGMENT_REFERENCE",
        "value": "WireElement_00064"
      }
    ],
    "type": "MODEL_REFERENCE"
  }
]